### ISSUE TITLE
blazor project / razor files

### ISSUE THREAD

**Issue by drma-tech** [meziantou/Meziantou.Analyzer#1306]
As someone who works with Blazor projects, I'm missing these validations in Razor files. Is this possible?

**Comment by meziantou** [comment]
To support razor files, you need to analyze generated code (the code generated by the razor source generator). An analyzer must explicitly opt-in. This means all generated code will start reporting issues (without control from the user). That will includes all source generators (blazor, regex, option binder, library imports, ...), and also files such as .g.cs or files having special comments.
The issue is that there is no possible selection from a user. The analyzer choose to analyze generated code or not.
That why I don't enable the flag yet except for rules that should clearly analyze generated code. If I stat enabling all rules, I'm not sure the experience will be great. Most code will break because the generator don't produce code that is valid for some MA rules.

**Comment by drma-tech** [comment]
I didn't understand half of what you said.

But... if there were a special configuration or parameter to enable validation for Razor files—at least temporarily...

**Comment by meziantou** [comment]
Let me explain the issue in simple terms ;)

When creating an analyzer (what this project does), you have two options:
- Don’t report issues in generated code
- Report issues in generated code

This is a static flag. You cannot dynamically enable it based on a user configuration in an editorconfig file.

Razor is compiled into C# code, which counts as generated code. To catch issues in Razor, you need to opt in to analyzing generated code.

The problem is, there’s a lot more generated code in a project, and there’s no way to say “analyze Razor-generated code but ignore other generated code.” Most of the time, you can’t control generated code, and it might not even exist on disk. If an analyzer flags a warning in that code, it could break your build with no way to fix it.

So, if I opt in to analyzing all generated code for all rules, there’s a high chance many projects won’t build anymore due to new errors or warnings.

It’s not just about adding a flag—it’s a limitation of Roslyn.

Maybe there would be a way to do it by re-implementing all the logic from Roslyn to determine if a file is generated or not. But currently out of scope for this project. I may reconsider it in the future, but I'm pretty sure it can also have performance impact as it needs to analyze more code.

**Comment by meziantou** [comment]
I've added a way to optin generated code using an environment variable https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/generated-code.md

```
MEZIANTOU_ANALYZER_GENERATED_CODE=1
```

**Comment by drma-tech** [comment]
Something just crossed my mind (I haven't tested it yet, so I don't know the actual scope).

My idea wasn't really to validate the C# code we put inside Razor (for that, you'd just move the code to a partial class, where it gets validated normally), but rather to validate the Razor code itself—for example, validating lambdas, parameters, etc.

**Comment by meziantou** [comment]
- C# partial files doesn't need the new environment variable to be analyzed. It's part of your code, so not a generated file.
- The new env var is only for the content of the razor file (`@code` or C# in the HTML section) + anything generated by the razor compiler.

**Comment by drma-tech** [comment]
So, it’s just as I suspected. It’s still only validating the C# code, not the "HTML" content. I’m refactoring some things in my projects—like removing unnecessary lambdas, reviewing parameters, component interactions, and so on—which is why I wondered if your component could handle that validation as well.

**Comment by meziantou** [comment]
It can analyze HTML code once it's converted to C#. So, we can analyze stuff like:

````
protected override void BuildRenderTree(RenderTreeBuilder __builder)
{
	__builder.OpenElement(0, "div");
	__builder.AddContent(1, Param);
	__builder.CloseElement();
	if (Param == 0)
	{
		__builder.OpenComponent<TestComponent>(2);
		__builder.AddComponentParameter(3, "Param", Microsoft.AspNetCore.Components.CompilerServices.RuntimeHelpers.TypeCheck(1));
		__builder.CloseComponent();
	}
}
````

This is what the Razor rules already do.
